include/EASTL/sort.h

```diff
@@ -205,77 +205,144 @@ namespace eastl
-	/// not the more flexible RandomAccessIterator, we implement the sort by doing a for loop within
-	/// a for loop. If we were to specialize this for a RandomAccessIterator, we could replace the
-	/// inner for loop with a call to upper_bound, which would be faster.
-	///
-	template <typename BidirectionalIterator, typename StrictWeakOrdering>
-	void insertion_sort(BidirectionalIterator first, BidirectionalIterator last, StrictWeakOrdering compare)
+	/* Insertion Sort
+	 *
+	 * An in-place and stable sorting algorithm that is fast for small arrays,
+	 * making it useful in algorithms like Quick Sort, Merge Sort, most-significant-
+	 * digit Radix Sorts, and Quick Select, which divide their inputs into smaller (semi)sorted segments.
+	 *
+	 * An improvement for larger inputs would be to use upper_bound to find the insertion position,
+	 * followed by memmove, but it will still be slow for large inputs, and not worth it for small ones.
+	 *
+	 * If base[-1] is a sentinel the short-circuit can be removed, but this is a general purpose version.
+	 */
+	
+	//compare functor version
+	template<class Iter, class Compare>//Bidirectional Iterator, though something like a list shouldn't use this
+	void insertion_sort(Iter const base, Iter const past, Compare cmp)
 	{
-		typedef typename eastl::iterator_traits<BidirectionalIterator>::value_type value_type;
-
-		if(first != last) // if the range is non-empty...
-		{
-			BidirectionalIterator iCurrent, iNext, iSorted = first;
-
-			for(++iSorted; iSorted != last; ++iSorted)
-			{
-				const value_type temp(*iSorted);
-
-				iNext = iCurrent = iSorted;
-
-				// Note: The following loop has a problem: it can decrement iCurrent to before 'first'.
-				// It doesn't dereference the iterator, but std STL disallows that operation. This isn't 
-				// a problem for EASTL containers and ranges, as they support a single decrement of first,
-				// but std STL iterators may have a problem with it. Dinkumware STL, for example, will assert.
-				// To do: Fix this loop to not decrement like so.
-				for(--iCurrent; (iNext != first) && compare(temp, *iCurrent); --iNext, --iCurrent)
-				{
-					EASTL_VALIDATE_COMPARE(!compare(*iCurrent, temp)); // Validate that the compare function is sane.
-					*iNext = *iCurrent;
-				}
-
-				*iNext = temp;
-			}
-		}
-	} // insertion_sort
-
-
-
-	template <typename BidirectionalIterator>
-	void insertion_sort(BidirectionalIterator first, BidirectionalIterator last)
+	    typedef typename eastl::iterator_traits<Iter>::value_type value_type;
+	
+	    if (base!=past)
+	    {   //this could perhaps be written more idiomatic: for (T *left=base, *sorted=base+1; sorted!=past; left=sorted++)
+	        for (Iter left=base, sorted=base; ++sorted!=past; left=sorted)
+	        {
+	            if (cmp(*sorted, *left))
+	            {
+	                const value_type val=*sorted;//should we care about doing moves instead?
```

Comment:
Was there a reason why you changed from 'const value_type temp(*iSorted);' an copy-assignment expression?  I believe in practice these should be equivalent as (I believe) the standard this will always be directed-initialization (ie. not a copy-constructed or copy-assigned).  I was curious if there was a reason for the change?  If there wasn't a problem you were working around I'd prefer to keep the behaviour as is to avoid any subtle knock-on affects.